schema.go

```diff
@@ -419,50 +437,74 @@ func (schema *Schema) ResolveSpaceIndex(s interface{}, i interface{}) (uint32, u
 		panic("unexpected type of space param")
 	}
 
-	if i != nil {
-		switch i := i.(type) {
-		case string:
-			if schema == nil {
-				return spaceNo, indexNo, fmt.Errorf("schema is not loaded")
-			}
-			if space == nil {
-				if space, ok = schema.SpacesById[spaceNo]; !ok {
-					return spaceNo, indexNo, fmt.Errorf("there is no space with id %d", spaceNo)
-				}
-			}
-			if index, ok = space.Indexes[i]; !ok {
-				err := fmt.Errorf("space %s has not index with name %s", space.Name, i)
-				return spaceNo, indexNo, err
-			}
-			indexNo = index.Id
-		case uint:
-			indexNo = uint32(i)
-		case uint64:
-			indexNo = uint32(i)
-		case uint32:
-			indexNo = i
-		case uint16:
-			indexNo = uint32(i)
-		case uint8:
-			indexNo = uint32(i)
-		case int:
-			indexNo = uint32(i)
-		case int64:
-			indexNo = uint32(i)
-		case int32:
-			indexNo = uint32(i)
-		case int16:
-			indexNo = uint32(i)
-		case int8:
-			indexNo = uint32(i)
-		case Index:
-			indexNo = i.Id
-		case *Index:
-			indexNo = i.Id
-		default:
-			panic("unexpected type of index param")
+	return spaceNo, nil
+}
+
+// ResolveIndex tries to resolve an index number.
+// Note: i can be a number, string, or an object of Index type.
+func (schema *Schema) ResolveIndex(i interface{}, spaceNo uint32) (uint32, error) {
+	var (
+		indexNo uint32
+		space   *Space
+		index   *Index
+		ok      bool
+	)
+
+	if i == nil {
+		return 0, nil
+	}
+	switch i := i.(type) {
+	case string:
+		if schema == nil {
+			return indexNo, fmt.Errorf("schema is not loaded")
+		}
+		if space, ok = schema.SpacesById[spaceNo]; !ok {
+			return indexNo, fmt.Errorf("there is no space with id %d", spaceNo)
 		}
+		if index, ok = space.Indexes[i]; !ok {
+			err := fmt.Errorf("space %s has not index with name %s", space.Name, i)
+			return indexNo, err
+		}
+		indexNo = index.Id
+	case uint:
+		indexNo = uint32(i)
+	case uint64:
+		indexNo = uint32(i)
+	case uint32:
+		indexNo = i
+	case uint16:
+		indexNo = uint32(i)
+	case uint8:
+		indexNo = uint32(i)
+	case int:
+		indexNo = uint32(i)
+	case int64:
+		indexNo = uint32(i)
+	case int32:
+		indexNo = uint32(i)
+	case int16:
+		indexNo = uint32(i)
+	case int8:
+		indexNo = uint32(i)
+	case Index:
+		indexNo = i.Id
+	case *Index:
+		indexNo = i.Id
+	default:
+		panic("unexpected type of index param")
 	}
 
-	return spaceNo, indexNo, nil
+	return indexNo, nil
+}
+
+func (schema *Schema) NamesUseSupported() bool {
+	return schema.SpaceAndIndexNamesSupported
+}
+
+type DefaultSchemaResolver struct {
+	Schema
```

Comment:
This is a strange decision, as a result of which we still have `if` which we wanted to get away from.

Reply:
There is no additional if:
```
func (schema *Schema) NamesUseSupported() bool {
	return schema.spaceAndIndexNamesSupported
}

type defaultSchemaResolver struct {
	Schema
}

func (*defaultSchemaResolver) NamesUseSupported() bool {
	return false
}

```

Reply:
Let's make a concrete implementation better.

Reply:
As a result, I got two identical types `Schema` and `defaultSchemaResolver`. They both have `ResolveSpace`, `ResolveIndex` and `NamesUseSupported` methods and  `spaceAndIndexNamesSupported` field. Is it ok to remove `defaultSchemaResolver` and, maybe use `Schema` with default values instead of `defaultSchemaResolver` variable? So the code will look like this:

connection.go:
```
func Connect(ctx context.Context, addr string, opts Opts) (conn *Connection, err error) {
	conn = &Connection{
		addr:             addr,
		requestId:        0,
		contextRequestId: 1,
		Greeting:         &Greeting{},
		control:          make(chan struct{}),
		opts:             opts.Clone(),
		dec:              msgpack.NewDecoder(&smallBuf{}),
		schemaResolver:   &Schema{},
	}

...

if err := pack(&shard.buf, shard.enc, reqid, req, streamId, conn.schemaResolver); err != nil {
```

schema.go:
```
func (conn *Connection) loadSchema() (err error) {
	schema := new(Schema)
	schema.SpacesById = make(map[uint32]*Space)
	schema.Spaces = make(map[string]*Space)

	checkNamesSupported(conn, schema)
	conn.schemaResolver.spaceAndIndexNamesSupported = schema.NamesUseSupported()

...

	conn.lockShards()
	conn.Schema = schema
	conn.schemaResolver = schema
	conn.unlockShards()
```

Reply:
It's a bit different implementations in fact:

```Go
func resolveSpaceNumber(s interface{}) (uint32, error) {
	var spaceNo uint32

	switch s := s.(type) {
	case uint:
		spaceNo = uint32(s)
	case uint64:
		spaceNo = uint32(s)
	case uint32:
		spaceNo = s
	case uint16:
		spaceNo = uint32(s)
	case uint8:
		spaceNo = uint32(s)
	case int:
		spaceNo = uint32(s)
	case int64:
		spaceNo = uint32(s)
	case int32:
		spaceNo = uint32(s)
	case int16:
		spaceNo = uint32(s)
	case int8:
		spaceNo = uint32(s)
	case Space:
		spaceNo = s.Id
	case *Space:
		spaceNo = s.Id
	default:
		panic("unexpected type of space param")
	}

	return spaceNo, nil
}

func resolveIndexNumber(i interface{}) (uint32, error) {
	var indexNo uint32

	switch i := i.(type) {
	case uint:
		indexNo = uint32(i)
	case uint64:
		indexNo = uint32(i)
	case uint32:
		indexNo = i
	case uint16:
		indexNo = uint32(i)
	case uint8:
		indexNo = uint32(i)
	case int:
		indexNo = uint32(i)
	case int64:
		indexNo = uint32(i)
	case int32:
		indexNo = uint32(i)
	case int16:
		indexNo = uint32(i)
	case int8:
		indexNo = uint32(i)
	case Index:
		indexNo = i.Id
	case *Index:
		indexNo = i.Id
	default:
		panic("unexpected type of index param")
	}

	return indexNo, nil
}

func (schema *Schema) ResolveSpace(s interface{}) (uint32, error) {
	if str, ok := s.(string); ok {
		space, ok := schema.Spaces[str]
		if !ok {
			return 0, fmt.Errorf("there is no space with name %s", s)
		}
		return space.Id, nil
	}
	return resolveSpaceNumber(s)
}

func (r defaultSchemaResolver) ResolveSpace(s interface{}) (uint32, error) {
	if _, ok := s.(string); ok {
		return 0, fmt.Errorf("schema is not loaded")
	}
	return resolveSpaceNumber(s)
}

// ResolveIndex tries to resolve an index number.
// Note: i can be a number, string, or an object of Index type.
func (schema *Schema) ResolveIndex(i interface{}, spaceNo uint32) (uint32, error) {
	if str, ok := i.(string); ok {
		space, ok := schema.SpacesById[spaceNo]
		if !ok {
			return 0, fmt.Errorf("there is no space with id %d", spaceNo)
		}
		if index, ok = space.Indexes[i]; !ok {
			err := fmt.Errorf("space %s has not index with name %s", space.Name, i)
			return indexNo, err
		}
		indexNo = index.Id

	}
	return resolveIndexNUmber(i)
}

func (r defaultSchemaResolver) ResolveIndex(s interface{}) (uint32, error) {
	if _, ok := i.(string); ok {
		return 0, fmt.Errorf("schema is not loaded")
	}
	return resolveSpaceNumber(s)
}
```

Reply:
Updated.